san/impl/derive_type.cpp

```diff
@@ -0,0 +1,256 @@
+        }
+        m_searchArray = false;
+        node->operand2->accept(*this);
+        if (!typeIsBase(node->operand2->type)) {
+            error(node->operand2->begin, "invalid type for array index");
+        }
+        node->type = m_arrayInnerType;
+    } else if (node->operation == lexer::TokenType::Dot) {
+        // if operation is dot ntation
+        // check that the first operand is a record -???
+        // take the type of key arg
+        m_searchFiled = true;
+        node->operand2->accept(*this);
+        m_searchFiled = false;
+        if (m_recordField == "") {
+            error(node->operand2->begin, "field name of record not found");
+        }
+        // search for the type of key arg
+        m_searchRecord = true;
+        node->operand1->accept(*this);
+        m_searchRecord = false;
+
+        if (m_recordInnerType == nullptr) {
+            error(node->operand1->begin,
+                  "invalid operation '.' on the given type");
+        }
+
+        node->type = m_recordInnerType;
+
+    } else if (node->operand1->type != nullptr &&
+               node->operand2->type != nullptr &&
+               node->operand1->type->getTypeKind() !=
+                   node->operand2->type->getTypeKind()) {
+        sPtr<ast::Type> type1 = node->operand1->type;
+        if (!typeIsBase(type1)) {
+            error(node->operand1->begin, "invalid type of expression");
+        }
+
+        sPtr<ast::Type> type2 = node->operand2->type;
+        if (!typeIsBase(type2)) {
+            error(node->operand2->begin, "invalid type of expression");
+        }
+
+        node->type = getGreaterType(type1, type2);
+    } else {
+        // types are equal
+        node->type = node->operand1->type;
+    }
+}
+
+void DeriveType::visit(Primary*) {}
+
+void DeriveType::visit(IntegerLiteral*) {}
+
+void DeriveType::visit(RealLiteral*) {}
+
+void DeriveType::visit(BooleanLiteral*) {}
+
+void DeriveType::visit(Identifier* node) {
+    node->variable->accept(*this);
+
+    if (m_searchFiled) {
+        m_recordField = node->name;
+    }
+    node->type = node->variable->type;
+}
+
+// already has a type
+void DeriveType::visit(RoutineCall*) {}
+
+sPtr<ast::Type> DeriveType::getGreaterType(sPtr<ast::Type> type1,
+                                           sPtr<ast::Type> type2) {
+    // int  & real = real
+    // bool & int  = int
+    // bool & real = real
+    if (type1->getTypeKind() == type2->getTypeKind()) {
+        return type1; // if equal
+    }
+    if (type1->getTypeKind() == ast::TypeKind::Boolean) {
+        return type2;
+    }
+    if (type2->getTypeKind() == ast::TypeKind::Boolean) {
+        return type1;
+    }
+    if (type1->getTypeKind() == ast::TypeKind::Integer) {
+        return type2; // return real
+    }
+    if (type2->getTypeKind() == ast::TypeKind::Integer) {
+        return type1; // return real
+    }
+    return type1;
+}
+bool DeriveType::typeIsBase(sPtr<ast::Type> type) {
```

Comment:
Wouldn't `primitive` be more descriptive than `base`?

Reply:
Also, you don't need to use any `ast::` qualifier since you have `using namespace ast;` in the beginning of the file